File: metadata/mill-metadata-core/src/test/java/io/qpointz/mill/metadata/service/JsonSchemaFacetContentValidatorTest.java
package io.qpointz.mill.metadata.service;

import io.qpointz.mill.metadata.domain.ValidationResult;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

class JsonSchemaFacetContentValidatorTest {

    private JsonSchemaFacetContentValidator validator;

    @BeforeEach
    void setUp() {
        validator = new JsonSchemaFacetContentValidator();
    }

    @Test
    void shouldPass_validData() {
        Map<String, Object> schema = Map.of(
                "type", "object",
                "properties", Map.of(
                        "name", Map.of("type", "string"),
                        "count", Map.of("type", "integer")
                ),
                "required", List.of("name")
        );

        Map<String, Object> data = Map.of("name", "test", "count", 42);

        ValidationResult result = validator.validate(schema, data);
        assertTrue(result.valid());
    }

    @Test
    void shouldFail_missingRequiredField() {
        Map<String, Object> schema = Map.of(
                "type", "object",
                "properties", Map.of(
                        "name", Map.of("type", "string")
                ),
                "required", List.of("name")
        );

        Map<String, Object> data = Map.of("count", 42);

        ValidationResult result = validator.validate(schema, data);
        assertFalse(result.valid());
        assertFalse(result.errors().isEmpty());
    }

    @Test
    void shouldFail_wrongType() {
        Map<String, Object> schema = Map.of(
                "type", "object",
                "properties", Map.of(
                        "count", Map.of("type", "integer")
                )
        );

        Map<String, Object> data = Map.of("count", "not-a-number");

        ValidationResult result = validator.validate(schema, data);
        assertFalse(result.valid());
    }

    @Test
    void shouldPass_noRequiredFields() {
        Map<String, Object> schema = Map.of(
                "type", "object",
                "properties", Map.of(
                        "name", Map.of("type", "string")
                )
        );

        Map<String, Object> data = Map.of();

        ValidationResult result = validator.validate(schema, data);
        assertTrue(result.valid());
    }
}

File: metadata/mill-metadata-core/src/test/kotlin/io/qpointz/mill/metadata/service/JsonSchemaFacetContentValidatorTest.kt
package io.qpointz.mill.metadata.service

import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

class JsonSchemaFacetContentValidatorTest {

    private lateinit var validator: JsonSchemaFacetContentValidator

    @BeforeEach
    fun setUp() {
        validator = JsonSchemaFacetContentValidator()
    }

    @Test
    fun shouldPass_validData() {
        val schema = mapOf<String, Any>(
            "type" to "object",
            "properties" to mapOf("name" to mapOf("type" to "string"), "count" to mapOf("type" to "integer")),
            "required" to listOf("name")
        )
        val result = validator.validate(schema, mapOf("name" to "test", "count" to 42))
        assertTrue(result.valid)
    }

    @Test
    fun shouldFail_missingRequiredField() {
        val schema = mapOf<String, Any>(
            "type" to "object",
            "properties" to mapOf("name" to mapOf("type" to "string")),
            "required" to listOf("name")
        )
        val result = validator.validate(schema, mapOf("count" to 42))
        assertFalse(result.valid)
        assertFalse(result.errors.isEmpty())
    }

    @Test
    fun shouldFail_wrongType() {
        val schema = mapOf<String, Any>(
            "type" to "object",
            "properties" to mapOf("count" to mapOf("type" to "integer"))
        )
        val result = validator.validate(schema, mapOf("count" to "not-a-number"))
        assertFalse(result.valid)
    }

    @Test
    fun shouldPass_noRequiredFields() {
        val schema = mapOf<String, Any>(
            "type" to "object",
            "properties" to mapOf("name" to mapOf("type" to "string"))
        )
        val result = validator.validate(schema, emptyMap<String, Any>())
        assertTrue(result.valid)
    }
}
